File: modules/react-native-studio-recorder/android/src/main/java/com/margelo/nitro/studiorecorder/StudioRecorderHybridObject.kt
package com.margelo.nitro.studiorecorder

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip

@DoNotStrip
@Keep
class StudioRecorderHybridObject : HybridNitroStudioRecorderSpec() {

  override val memorySize: Long
    get() = 0L

  override var isRecording: Boolean = false
    private set

  private var recordingStartMs: Long = 0L

  override fun startRecording(onMeterUpdate: (db: Double) -> Unit) {
    // TODO (Level 2 Android):
    // 1) Configure/start MediaRecorder
    // 2) Emit dB values on an interval from maxAmplitude
    // 3) Save temp file path + recording id
    if (isRecording) return
    isRecording = true
    recordingStartMs = System.currentTimeMillis()
    onMeterUpdate(-45.0)
  }

  override fun stopRecording(): RecordingResult {
    // TODO (Level 2 Android):
    // 1) Stop/release MediaRecorder
    // 2) Return id/filePath/durationMs
    // 3) Persist metadata (Level 3)
    val durationMs =
      if (isRecording) (System.currentTimeMillis() - recordingStartMs).toDouble() else 0.0

    isRecording = false
    recordingStartMs = 0L

    return RecordingResult("stub", "", durationMs)
  }

  override fun cancelRecording() {
    // TODO (Level 2 Android): stop/release and remove temp file.
    isRecording = false
    recordingStartMs = 0L
  }

  override fun getRecordings(): Array<RecordingInfo> {
    // TODO (Level 3 Android): load persisted metadata list.
    return emptyArray()
  }

  override fun deleteRecording(id: String) {
    // TODO (Level 3 Android): delete file + remove metadata entry.
  }
}

File: modules/react-native-studio-recorder/ios/StudioRecorderHybridObject.swift
import AVFoundation
import Foundation
import NitroModules

/// StudioRecorderHybridObject.swift
///
/// TODO (Levels 2-3): Implement the iOS Nitro hybrid object.
///
/// Suggested split:
/// - Level 2: recording lifecycle + live metering callback
/// - Level 3: persistence (list/delete metadata + files)
class StudioRecorderHybridObject: HybridNitroStudioRecorderSpec {

  var memorySize: Int { 0 }

  private(set) var isRecording: Bool = false
  private var recordingStartDate: Date?

  func startRecording(onMeterUpdate: @escaping (_ db: Double) -> Void) throws {
    // TODO (Level 2 iOS):
    // 1) Check/request microphone permission
    // 2) Configure AVAudioSession
    // 3) Start AVAudioEngine + tap inputNode
    // 4) Stream dB values via onMeterUpdate callback
    guard !isRecording else { return }
    isRecording = true
    recordingStartDate = Date()
    onMeterUpdate(-45)
  }

  func stopRecording() throws -> RecordingResult {
    // TODO (Level 2 iOS):
    // 1) Stop engine, remove tap, close file
    // 2) Return id/filePath/durationMs
    // 3) Persist metadata entry (Level 3)
    guard isRecording else {
      throw NSError(
        domain: "StudioRecorder",
        code: 2,
        userInfo: [NSLocalizedDescriptionKey: "Not currently recording."]
      )
    }

    isRecording = false
    let durationMs = Double(Date().timeIntervalSince(recordingStartDate ?? Date()) * 1000)
    recordingStartDate = nil

    return RecordingResult(
      id: "stub",
      filePath: "",
      durationMs: durationMs
    )
  }

  func cancelRecording() throws {
    // TODO (Level 2 iOS): stop engine and delete in-progress temp file.
    isRecording = false
    recordingStartDate = nil
  }

  func getRecordings() throws -> [RecordingInfo] {
    // TODO (Level 3 iOS): load persisted metadata (sorted newest first).
    return []
  }

  func deleteRecording(id: String) throws {
    // TODO (Level 3 iOS): delete file + remove metadata entry.
  }
}
